src/veritysetup/veritysetup.c

```diff
@@ -274,6 +277,95 @@ static int parse_options(const char *options) {
         return r;
 }
 
+static int find_dps_signature(
+                const char *data_device,
+                void *root_hash,
+                size_t root_hash_size,
+                char **ret_root_hash_signature) {
+
+        _cleanup_(loop_device_unrefp) LoopDevice *loop_device = NULL;
+        _cleanup_(dissected_image_unrefp) DissectedImage *dissected_image = NULL;
+        _cleanup_(verity_settings_done) VeritySettings verity = VERITY_SETTINGS_DEFAULT;
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        sd_device *parent = NULL;
+        _cleanup_free_ char *decoded_signature = NULL, *encoded_signature = NULL;
+        const char *block_device_path;
+        DissectImageFlags dissect_image_flags =
+                DISSECT_IMAGE_GPT_ONLY |
+                DISSECT_IMAGE_USR_NO_ROOT |
+                DISSECT_IMAGE_ADD_PARTITION_DEVICES |
+                DISSECT_IMAGE_DEVICE_READ_ONLY;
+        ssize_t len;
+        int r;
+
+        assert(data_device);
+        assert(root_hash);
+        assert(root_hash_size > 0);
+        assert(ret_root_hash_signature);
+
+        if (!startswith(data_device, "/dev/"))
+                return -EINVAL;
+
+        verity.root_hash_size = root_hash_size;
+        verity.root_hash = malloc(root_hash_size);
+        if (!verity.root_hash)
+                return log_oom_debug();
+        memcpy(verity.root_hash, root_hash, root_hash_size);
+
+        r = sd_device_new_from_devname(&device, data_device);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get udev device for data device: %m");
+
+        r = sd_device_get_parent(device, &parent);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get parent device for data device: %m");
+
+        r = sd_device_get_devname(parent, &block_device_path);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get block device path for data device: %m");
```

Comment:
`block_device_new_from_path()` with `BLOCK_DEVICE_LOOKUP_WHOLE_DISK` can be used.